airflow/utils/decorators.py

```diff
@@ -30,6 +30,9 @@
 
 
 def apply_defaults(func):
+
+    sig_cache = dict()
```

Comment:
Is this in the right place to cache more than a single function call?

```
@apply_defaults
def a():
    pass

@apply_defaults
def b():
    pass

```

I _think_ that `a` and `b` get a different sig_cache. Is that okay? expected? desired? behaviour

Reply:
I _think_ this is okay, since @apply_default is only applied to `__init__`.

It is also consistent with prior the change (i.e. a and b get different sig_cache)

Reply:
yea i think this should be the expected and desired result.
expected and desired b/c like @jgao54 mentioned, each __init__ *should* have been set up with its own wrapper as was the previous behavior.

It is required to be here but usage of the dict is slightly disingenuous. It's not here to actually cache more than a single function but to solve the closure problem of the `wrapper` function which refers to the out of scope sig_cache. If we did something like `sig_cache = (sig, non_optional_args)` this would merely rebind the sig_cache inside `wrapper`'s scope and not actually work at all.

if the dict is too confusing, should I change it to a private class with a private class instead?
something along the lines of:
```
def apply_defaults(func):
    class SignatureCache(object):
        def __init__(self):
            self.value = None
     
    sig_cache = SignatureCache()
... 
    def wrapper(...
...
         if sig_cache.value :
+            (sig, non_optional_args) = sig_cache.value 
+        else:
+            sig = signature(func)
+            non_optional_args = {
+                name for (name, param) in sig.parameters.items()
+                if param.default == param.empty and
+                param.name != 'self' and
+                param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)}
+            sig_cache.value= (sig, non_optional_args)
...

Would that be preferred?

Reply:
+1 on the suggestion, and may be worth commenting inline as you have explained above :)